--- src/eldap/eldap.erl.~1~	2012-04-26 16:04:16.000000000 +0200
+++ src/eldap/eldap.erl	2012-07-30 19:28:33.182004486 +0200
@@ -984,10 +984,10 @@
     ?INFO_MSG("LDAP connection on ~s:~p", [Host, S#eldap.port]),
     Opts = if S#eldap.tls == tls ->
                    [{packet, asn1}, {active, true}, {keepalive, true},
-                    binary | S#eldap.tls_options];
+                    inet6, binary | S#eldap.tls_options];
               true ->
                    [{packet, asn1}, {active, true}, {keepalive, true},
-                    {send_timeout, ?SEND_TIMEOUT}, binary]
+                    {send_timeout, ?SEND_TIMEOUT}, inet6, binary]
            end,
     SocketData = case S#eldap.tls of
 		     tls ->
