commit 1211c347c5a574739616ccadf4b3b54686f9051f
Author: Jonathan del Strother <jon.delStrother@audioboo.fm>
Date:   Mon Jan 12 08:19:38 2015 +0000

    Fix mocking belongs_to associations in Rails 4.2+

diff --git a/lib/rspec/active_model/mocks/mocks.rb b/lib/rspec/active_model/mocks/mocks.rb
index b67a9a0..fab2637 100644
--- a/lib/rspec/active_model/mocks/mocks.rb
+++ b/lib/rspec/active_model/mocks/mocks.rb
@@ -52,6 +52,10 @@ module RSpec::ActiveModel::Mocks
         send(key)
       end
 
+      # Rails>4.2 uses _read_attribute internally, as an optimized
+      # alternative to record['id']
+      alias_method :_read_attribute, :[]
+
       # Returns the opposite of `persisted?`
       def new_record?
         !persisted?
