diff -uNr brother2-1.0.0.orig/backend/brother.h brother2-1.0.0/backend/brother.h
--- brother2-1.0.0.orig/backend/brother.h	2005-05-12 01:27:49.000000000 -0400
+++ brother2-1.0.0/backend/brother.h	2006-01-08 03:11:08.000000000 -0500
@@ -294,7 +294,7 @@
 #define PSIZE_HMARGIN     20			// 左右マージン(0.1mm単位)
 
 #define MAX_PATH 256
-#define BROTHER_SANE_DIR "/usr/local/Brother/sane/"
+#define BROTHER_SANE_DIR "/usr/share/sane/Brother/"
 #define BROTHER_GRAYCMDATA_DIR "GrayCmData/"
 
 #endif
