--- lib/rsi/query.rb.~1~	2011-10-05 10:42:00.898503030 +0200
+++ lib/rsi/query.rb	2011-10-05 10:43:05.844581005 +0200
@@ -20,10 +20,10 @@
 
   class ANDQuery < Query
     def evaluate( locator )
-      ret_set = nil
+      ret_set = []
       @subqueries.each do |q|
         set = q.evaluate( locator )
-        if ret_set.nil?
+        if ret_set.empty?
           ret_set = set
         else
           ret_set = ret_set & set
