From f121814718c9907b20058dc9357b80a53afab821 Mon Sep 17 00:00:00 2001
From: Matt Brictson <matt@mattbrictson.com>
Date: Sat, 2 Jan 2021 13:52:30 -0800
Subject: [PATCH] Fix "undefined method safe_append=" on Ruby 3.0.0

---
 lib/xray/engine.rb | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/xray/engine.rb b/lib/xray/engine.rb
index 6f9024a..fc55bd7 100644
--- a/lib/xray/engine.rb
+++ b/lib/xray/engine.rb
@@ -21,10 +21,10 @@ class Engine < ::Rails::Engine
       ActionView::Template.class_eval do
         extend Xray::Aliasing
 
-        def render_with_xray(*args, &block)
+        def render_with_xray(*args, **kwargs, &block)
           path = identifier
           view = args.first
-          source = render_without_xray(*args, &block)
+          source = render_without_xray(*args, **kwargs, &block)
 
           suitable_template = !(view.respond_to?(:mailer) && view.mailer) &&
                               !path.include?('_xray_bar') &&
