--- Rakefile.orig	2010-04-22 16:30:22.965914139 +0200
+++ Rakefile	2010-04-22 16:30:35.397914130 +0200
@@ -1,25 +1,5 @@
 require 'rake'
 
-require 'jeweler'
-
-Jeweler::Tasks.new do |s|
-  s.name = "rest-client"
-  s.description = "A simple HTTP and REST client for Ruby, inspired by the Sinatra microframework style of specifying actions: get, put, post, delete."
-  s.summary = "Simple HTTP and REST client for Ruby, inspired by microframework syntax for specifying actions."
-  s.author = "Adam Wiggins"
-  s.email = "rest.client@librelist.com"
-  s.homepage = "http://github.com/archiloque/rest-client"
-  s.rubyforge_project = "rest-client"
-  s.has_rdoc = true
-  s.files = FileList["[A-Z]*", "{bin,lib,spec}/**/*"]
-  s.executables = %w(restclient)
-  s.add_dependency("mime-types", ">= 1.16")
-  s.add_development_dependency("webmock", ">= 0.9.1")
-  s.add_development_dependency("rspec")
-end
-
-Jeweler::RubyforgeTasks.new
-
 ############################
 
 require 'spec/rake/spectask'
