--- bindings/ruby/extconf.rb.~1~	2008-11-14 13:41:14.000000000 +0100
+++ bindings/ruby/extconf.rb	2009-09-08 19:36:23.752462307 +0200
@@ -1,8 +1,8 @@
 require 'mkmf'
-dir_config("senna", `senna-cfg --prefix`.chomp)
+dir_config("senna", `../../senna-cfg --prefix`.chomp)
 #$LOCAL_LIBS << '  -lsenna -lmecab -lstdc++'
-$LOCAL_LIBS << ' ' + `senna-cfg --libs`.chomp
-$CFLAGS << ' ' + `senna-cfg --cflags`.chomp
+$LOCAL_LIBS << ' ' + '-L../../lib/.libs ' + `../../senna-cfg --libs`.chomp
+$CFLAGS << ' ' + '-I../.. ' + `../../senna-cfg --cflags`.chomp
 if have_header("senna.h") and have_library("senna", "sen_init")
 #  system "swig -ruby senna_api.i"
   create_makefile("senna_api")
