--- lib/responds_to_parent/action_controller.rb.~1~	2012-09-19 17:10:06.937785043 +0200
+++ lib/responds_to_parent/action_controller.rb	2012-09-19 17:13:19.220322750 +0200
@@ -20,6 +20,7 @@
         response.headers['Content-Type'] = 'text/html; charset=UTF-8'
 
         # Clear out the previous render to prevent double render
+        @_response_body = nil
         response.request.env['action_controller.instance'].instance_variable_set(:@_response_body, nil)
 
         # Eval in parent scope and replace document location of this frame 
