--- original/libxml-ruby-1.1.3/ext/libxml/ruby_xml_reader.c	2009-08-04 14:13:59.000000000 -0700
+++ libxml-ruby-1.1.3/ext/libxml/ruby_xml_reader.c	2009-08-04 14:59:11.000000000 -0700
@@ -747,6 +747,7 @@ static VALUE rxml_reader_attribute(VALUE
 {
   xmlTextReaderPtr reader;
   xmlChar *attr;
+  VALUE result = Qnil;
 
   reader = rxml_text_reader_get(self);
 
@@ -758,7 +759,13 @@ static VALUE rxml_reader_attribute(VALUE
   {
     attr = xmlTextReaderGetAttribute(reader, (const xmlChar *) StringValueCStr(key));
   }
-  return (attr == NULL ? Qnil : rb_str_new2((const char*)attr));
+
+  if (attr != NULL)
+  {
+    result = rb_str_new2((const char*) attr);
+    xmlFree(attr);
+  }
+  return result;
 }
 
 /*
