diff --git a/lib/bootstrap_form/form_group_builder.rb b/lib/bootstrap_form/form_group_builder.rb
index 6bd733e..c173d76 100644
--- a/lib/bootstrap_form/form_group_builder.rb
+++ b/lib/bootstrap_form/form_group_builder.rb
@@ -28,7 +28,7 @@ module BootstrapForm
     def form_group_builder_options(options, method)
       options.symbolize_keys!
       options = convert_form_tag_options(method, options) if acts_like_form_tag
-      options[:required] = form_group_required(options) if !options[:skip_label] && options.key?(:skip_required)
+      options[:required] = form_group_required(options) if !options[:skip_label]
       options
     end
 
