diff --git a/lib/will_paginate/active_record.rb b/lib/will_paginate/active_record.rb
index 5fb2ece..3ab973a 100644
--- a/lib/will_paginate/active_record.rb
+++ b/lib/will_paginate/active_record.rb
@@ -2,10 +2,8 @@ require 'will_paginate/per_page'
 require 'will_paginate/page_number'
 require 'will_paginate/collection'
 require 'active_record'
-begin
+if ActiveRecord::VERSION::MAJOR >= 4
   require 'active_record/deprecated_finders'
-rescue LoadError
-  # only for Rails 4.1
 end
 
 module WillPaginate
